fil_fungible_token/src/receiver/mod.rs

```diff
@@ -1 +1,56 @@
+use fvm_ipld_encoding::RawBytes;
+use fvm_shared::{address::Address, econ::TokenAmount, error::ExitCode};
+use num_traits::Zero;
+use types::TokensReceivedParams;
+
+use crate::runtime::messaging::{Messaging, RECEIVER_HOOK_METHOD_NUM};
+use crate::token::TokenError;
+
 pub mod types;
+
+#[derive(Debug)]
+pub struct ReceiverHookGuard<T> {
+    address: Address,
+    params: Option<TokensReceivedParams>,
+    return_value: Option<T>,
```

Comment:
I don't think threading this otherwise-unrelated value here is good. I see how in the case that calling the hook is exactly the last thing the caller does that it can make for a concise syntax, but that's also potentially confusing about what's being returned.

Focus this type just on the hook thunk. The mint/transfer methods can return an anonymous tuple of the result and this hook thunk.

Reply:
I'll remove it then. It will look a little more awkward in use:
```Rust
let (mut hook, ret) = token.mint(...)?;
```
But it does give you access to the return value immediately.

The original concept was that the operation and receiver hook call should be seen as two halves of the one transaction, so the final result was returned by the hook call.